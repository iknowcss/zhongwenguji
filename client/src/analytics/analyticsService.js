import getConfig from '../getConfig';

export const initStateEnum = {
  UNINITIALIZED: 'UNINITIALIZED',
  INITIALIZED: 'INITIALIZED',
  WONT_INITIALIZE: 'WONT_INITIALIZE'
};

export default {
  state: initStateEnum.UNINITIALIZED,

  gtag() { /* noop */ },

  init() {
    if (this.state !== initStateEnum.UNINITIALIZED) {
      return;
    }
    if (getConfig().enableGA) {
      this.state = initStateEnum.INITIALIZED;
      console.log('GA Enabled');
      const gaScript = document.createElement('script');
      gaScript.setAttribute('async', 'true');
      gaScript.setAttribute('src', 'https://www.googletagmanager.com/gtag/js?id=UA-131912391-1');
      document.head.appendChild(gaScript);

      window.dataLayer = window.dataLayer || [];
      this.gtag = function () { window.dataLayer.push(arguments); };
      this.gtag('js', new Date());
      this.gtag('config', 'UA-131912391-1');
    } else {
      this.state = initStateEnum.WONT_INITIALIZE;
      console.log('GA Disabled');
    }
  },

  sendEvent(eventName, options) {
    if (this.state === initStateEnum.INITIALIZED) {
      this.gtag('event', eventName, options);
    }
  }
}
