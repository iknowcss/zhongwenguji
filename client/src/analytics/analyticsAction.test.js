import analyticsService from './analyticsService';
import {
  firstSwipe,
  completeTestAfterDuration,
  receiveKnownEstimate,
  reviewMissed
} from './analyticsAction';

describe('analyticsAction', () => {
  let sendEventStub;

  beforeEach(() => {
    sendEventStub = jest.spyOn(analyticsService, 'sendEvent');
  });

  afterEach(() => {
    sendEventStub.mockRestore();
  });

  it('firstSwipe', () => {
    firstSwipe('traditional');
    expect(sendEventStub).toHaveBeenCalledTimes(1);
    expect(sendEventStub.mock.calls[0]).toMatchSnapshot();
  });

  it('completeTestAfterDuration', () => {
    completeTestAfterDuration(42);
    expect(sendEventStub).toHaveBeenCalledTimes(1);
    expect(sendEventStub.mock.calls[0]).toMatchSnapshot();
  });

  it('receiveKnownEstimate', () => {
    receiveKnownEstimate(980);
    expect(sendEventStub).toHaveBeenCalledTimes(1);
    expect(sendEventStub.mock.calls[0]).toMatchSnapshot();
  });

  it('reviewMissed', () => {
    reviewMissed();
    expect(sendEventStub).toHaveBeenCalledTimes(1);
    expect(sendEventStub.mock.calls[0]).toMatchSnapshot();
  });
});
