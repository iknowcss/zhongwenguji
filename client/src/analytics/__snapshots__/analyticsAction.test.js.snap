// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyticsAction completeTestAfterDuration 1`] = `
Array [
  "completeTestAfterDuration",
  Object {
    "event_category": "Test",
    "value": 42,
  },
]
`;

exports[`analyticsAction firstSwipe 1`] = `
Array [
  "firstSwipe",
  Object {
    "event_category": "Test",
    "event_label": "traditional",
  },
]
`;

exports[`analyticsAction receiveKnownEstimate 1`] = `
Array [
  "receiveKnownEstimate",
  Object {
    "event_category": "Test results",
    "value": 980,
  },
]
`;

exports[`analyticsAction reviewMissed 1`] = `
Array [
  "reviewMissed",
  Object {
    "event_category": "Test results",
  },
]
`;
